wagl/reflectance.py

```diff
@@ -22,6 +24,502 @@
+    if outds is None:
+        outds = numpy.zeros(dims, dtype='float32')
+
+    # process by tile
+    for tile in acquisition.tiles():
+        # read the data corresponding to the current tile for all dataset
+        # the original f90 routine specified single precision
+        rad = acquisition.radiance_data(window=tile, out_no_data=NAN,
+                                        esun=esun).astype('float32')
+
+        # define the data mask (identify NaN pixels)
+        data_mask = ~numpy.isfinite(rad)
+
+        # lambertian
+        lambt = lambertian_block(rad, a[tile], b[tile], s[tile])
+
+        # account for original nulls and any evaluated nan's
+        lambt[data_mask] = NAN
+
+        outds[tile] = lambt
+
+    if isinstance(outds, numpy.ndarrray):
+        return outds
+
+
+def average_lambertian(acquisition, a, b, s, psf_kernel, esun=None,
+                       normalise=True):
+    """
+    Calculate the lambertian reflectance by taking into account the
+    the surrounding pixels.
+    Essentially we convolve the point spread function as a kernel
+    across the lambertian reflectance to get the surrounding
+    contribution.
+    Prior to convolution, the data is first smoothed by filling
+    null values with an average. Currently this null filling process
+    is evaluated by a run length average.
+    After convolution and before returning, nulls are re-inserted back
+    into the result.
+
+    TODO:
+        * Include an option to run convolution via fourier, to enable
+          faster processing for large kernels.
+        * Incorporate astropy for convolution to enable handling of
+          null data as well as NaN's.
+
+    :param acquisition:
+        This process was designed to be applied to lambertian
+        reflectance, but in practice it could be any 2D NumPy array.
+        Currently, the process is assuming null data to be -999.
+
+    :param psf_kernel:
+        A 2D kernel/filter representing the point spread function of
+        the atmosphere to resolve the atmospheric adjacency.
+
+    :param normalise:
+        A boolean indicating whether or not to normalise the
+        psf_kernel. Default is True.
+
+    :return:
+        A 2D NumPy array of type float32, with NaN's populating
+        invalid elements.
+    """
+    # normalise the kernel or not
+    if normalise:
+        psf_kernel = psf_kernel / psf_kernel.sum()
+
+    data = lambertian_tiled(acquisition, a, b, s, esun)
+    data_mask = ~numpy.isfinite(data)  # locate NaN's
+
+    # can we correctly apply row-length averages?
+    _, start_idx, end_idx = _sequential_valid_rows(data_mask)  # ignore all_valid for time being
+
+    # fill nulls with run-length averages
+    # _fill_nulls(data[start_idx:end_idx], null_mask[start_idx:end_idx])
+    _fill_nulls(data, data_mask)
+
+    # apply convolution
+    result = numpy.full(data.shape, fill_value=numpy.nan, dtype='float32')
+    ndimage.convolve(data[start_idx:end_idx].T, psf_kernel,
```

Comment:
in previous commit, convolution was done on radiance, but it is fixed here, did we test with Fuqins result after this commit?

Reply:
have now.  result is different from our first test, but still different from Fuqin's sample dataset.  but there is a lot of uncertainty around that sample data; how it was produced, code used etc.